### ISSUE TITLE
detect --source doesn't allow ~/ (tilde slash)

### ISSUE THREAD

**Issue by minispooner** [gitleaks/gitleaks#935]
**Describe the bug**
`detect --source` doesn't allow ~/ (tilde slash)

**To Reproduce**
`./gitleaks detect --source ~/cloned_repo`

**Expected behavior**
source param should be able to read `~` as home dir. it does work when I add absolute path though as a temp work around.

**Screenshots**
N/A. Error msg = "[TIME] FTL stat ~/cloned_repo: no such file or directory"

**Basic Info (please complete the following information):**
 - OS: ubuntu 22.04
 - Gitleaks Version: v8.9.0

**Additional context**

cc @zricethezav

**Comment by zricethezav** [comment]
@minispooner fair, I'm happy to accept contributions on this one

**Comment by maltemorgenstern** [comment]
@minispooner are you running gitleaks using the docker container?

I did some testing locally and can only partially reproduce this issue:
- When running gitleaks inside Debian 10 WSL the `~` gets expanded without any issues and gitleaks scans the home directory
  `./gitleaks detect --log-level debug -s ~/secrets --no-git`
- When running gitleaks __manually__ inside the docker container the `~` gets expanded without any issues and gitleaks scans the home directory
  ```
  docker run -it -v "${pwd}:/home/gitleaks/secrets" --entrypoint="" zricethezav/gitleaks:v8.13.0 sh
  $ gitleaks detect --log-level debug -s ~/secrets --no-git
  ```
- When running gitleaks __automatically__ inside the docker container the `~` does not get expanded and gitleaks fails
  ```
  docker run -it -v "${pwd}:/home/gitleaks/secrets" zricethezav/gitleaks:v8.13.0 detect --log-level debug -s ~/secrets --no-git
  FTL stat ~/secrets: no such file or directory
  ```

I added some debug logging and discovered that in the problematic case the `~` is not being expanded - while in the other cases gitleaks always gets an expanded source from the bash/shell.

```
# added logging here: https://github.com/zricethezav/gitleaks/blob/v8.13.0/cmd/root.go#L97

# working scenarios
DBG Extracted source in cmd/root: /home/gitleaks/secrets

# not working scenarios
DBG Extracted source in cmd/root: ~/secrets
```
Afterwards the `os.Stat` fails: https://github.com/zricethezav/gitleaks/blob/v8.13.0/cmd/root.go#L101


Looking into the Dockerfile reference this behavior seems to be documented: https://docs.docker.com/engine/reference/builder/#exec-form-entrypoint-example

> Unlike the shell form, the exec form does not invoke a command shell. This means that normal shell processing does not happen. For example, ENTRYPOINT [ "echo", "$HOME" ] will not do variable substitution on $HOME.


To sum it up:
- `~` should work fine when using a shell/bash (inside Debian, Ubuntu, ...)
- This will not work when passing it directly to the Dockerfile
  - But this should not be necessary because one can mount a folder to any location inside the container

@zricethezav I think this can be closed - WDYT?

**Comment by minispooner** [comment]
iirc, I came across the issue while running a bash loop over a hosts file in ubuntu 20.04 - not in docker. sorry I probably can't look into this one deeper. also im not familiar with go

**Comment by zricethezav** [comment]
> To sum it up:
> ~ should work fine when using a shell/bash (inside Debian, Ubuntu, ...)
> This will not work when passing it directly to the Dockerfile
> But this should not be necessary because one can mount a folder to any location inside the container

@maltemorgenstern I'm late to the game, but I agree with this. Closing. Thanks for investigating!
